src/routes/users.rs

```diff
@@ -54,6 +92,160 @@ async fn post_login(
     Ok(Redirect::to(&redirect_to).into_response())
 }
 
+async fn get_login_google(
+    State(state): State<AppState>,
+    session: Session,
+) -> ResponseResult<Response> {
+
+    let google_client_id = ClientId::new(
+        state.oauth_google_client_id
+    );
+    let google_client_secret = ClientSecret::new(
+        state.oauth_google_client_secret
+    );
+    let issuer_url = 
+        IssuerUrl::new("https://accounts.google.com".to_string()).context("failed to reach issuer")?;
+
+    let provider_metadata = task::spawn_blocking(move || {
+        GoogleProviderMetadata::discover(&issuer_url, http_client)
+        .context("failed to discover provider").unwrap()
+    }).await.context("failed to discover provider")?;
+    
+    let revocation_endpoint = provider_metadata
+        .additional_metadata()
+        .revocation_endpoint
+        .clone();
+    // Set up the config for the Google OAuth2 process.
+    let client = CoreClient::from_provider_metadata(
+        provider_metadata,
+        google_client_id,
+        Some(google_client_secret),
+    )
+    .set_redirect_uri(
+        RedirectUrl::new(state.base_url.to_string() + "/login_google_handler").expect("Invalid redirect URL"),
+    )
+    .set_revocation_uri(
+        RevocationUrl::new(revocation_endpoint).expect("Invalid revocation endpoint URL"),
+    );
+
+    // Generate the authorization URL to which we'll redirect the user.
+    let (authorize_url, csrf_state, nonce) = client
+        .authorize_url(
+            AuthenticationFlow::<CoreResponseType>::AuthorizationCode,
+            CsrfToken::new_random,
+            Nonce::new_random,
+        )
+        .add_scope(Scope::new("email".to_string()))
+        .add_scope(Scope::new("profile".to_string())).url();
+
+    // TODO: Store the CSRF and none states in a way that is more secure than this, although the current method is already quire secure.
```

Comment:
How would you store this more securely?

Reply:
I think there should be some algorithm that is made to encrypt the csrf_token and nonce values instead of directly storing to the session, for instance, we could hash the csrf_token prior to inserting it to the session and encrypt/bycrypt the nonce value